### PR TITLE
feat(ai-function): ai create testcase by groups and and export in fileType xmind

### PR DESCRIPTION
**Pull request by wangzhuzhen**
#### What this PR does / why we need it:
Optimize AI create testcase by groups from requirements and export in fileType xmind

#### Specified Reviewers:

/assign @sfwn

#### ChangeLog
Feature: Optimize AI create testcase by groups from requirements and export in fileType xmind（优化了 AI 测试用例根据需求内容进行分组生成，并支持导出 xmind 文件或保存）
| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English |    Feature: Optimize AI create testcase by groups from requirements and export in fileType xmind       |
| 🇨🇳 中文    |     优化了 AI 测试用例根据需求内容进行分组生成，并支持导出 xmind 文件或保存    |

#### Need cherry-pick to release versions?
/cherry-pick release/2.4

### CODE REVIEW

File: internal/apps/dop/endpoints/endpoints.go [hunk 1]
```diff
@@ -279,6 +279,7 @@ func (e *Endpoints) Routes() []httpserver.Endpoint {
 		{Path: "/api/testcases/actions/batch-copy", Method: http.MethodPost, Handler: e.BatchCopyTestCases},
 		{Path: "/api/testcases/actions/batch-clean-from-recycle-bin", Method: http.MethodDelete, Handler: e.BatchCleanTestCasesFromRecycleBin},
 		{Path: "/api/testcases/actions/export", Method: http.MethodGet, Handler: e.ExportTestCases},
+		{Path: "/api/testcases/actions/export", Method: http.MethodPost, Handler: e.ExportAIGeneratedTestCases},
```
Comment (sfwn) [inline internal/apps/dop/endpoints/endpoints.go:282]: Bad design.
Use `Get` for export. Maybe you should explicitly declare it's AI related on the url path.

File: internal/apps/dop/endpoints/testcase.go [hunk 2]
```diff
@@ -429,3 +433,88 @@ func (e *Endpoints) ImportTestCases(ctx context.Context, r *http.Request, vars m
...
+
+// ExportTestCases 导出测试用例
+func (e *Endpoints) ExportAIGeneratedTestCases(ctx context.Context, r *http.Request, vars map[string]string) (httpserver.Responser, error) {
+	identityInfo, err := user.GetIdentityInfo(r)
+	if err != nil {
+		return apierrors.ErrExportTestCases.NotLogin().ToResp(), nil
+	}
+
+	var pbReq pb.TestCaseExportRequest
+	//var req pb.TestCaseExportRequest
```
Comment (sfwn) [inline internal/apps/dop/endpoints/testcase.go:445]: remove useless code

File: internal/apps/dop/dao/file_records.go [hunk 3]
```diff
@@ -90,6 +90,11 @@ func (client *DBClient) GetRecord(id uint64) (*TestFileRecord, error) {
 	return &res, nil
 }
 
+// Delete Record by id
+func (client *DBClient) DeleteRecordBy(id uint64) error {
+	return client.Where("id =?", id).Delete(&TestFileRecord{}).Error
```
Comment (sfwn) [inline internal/apps/dop/dao/file_records.go:95]: format sql segment.

File: internal/core/openapi/legacy/api/apis/testplatform/testcase/export_ai_generated.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+package testcase
+
+import (
+	"net/http"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/internal/core/openapi/legacy/api/apis"
+)
+
+var EXPORT_AI_GENERATED_TESTCASES = apis.ApiSpec{
```
Comment (sfwn) [inline internal/core/openapi/legacy/api/apis/testplatform/testcase/export_ai_generated.go:24]: use protobuf only.

File: internal/pkg/ai-functions/functions/test-case/group.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/bundle"
+	"github.com/erda-project/erda/internal/apps/ai-proxy/vars"
+	"github.com/erda-project/erda/internal/pkg/ai-functions/sdk"
+	"github.com/erda-project/erda/pkg/common/apis"
+)
+
+type RequireMentGroup struct {
```
Comment (sfwn) [inline internal/pkg/ai-functions/functions/test-case/group.go:38]: Requirement

File: internal/pkg/ai-functions/functions/test-case/group.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	ID     int64
+	Groups []string
+}
+
+const (
+	EnvChatGPT4AIProxyModelId = "AI_PROXY_CHATGPT4_MODEL_ID"
```
Comment (sfwn) [inline internal/pkg/ai-functions/functions/test-case/group.go:44]: declared at provider config.

File: internal/pkg/ai-functions/functions/test-case/group.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+	for _, include := range issueRelations.IssueInclude {
+		if include.Type == apistructs.IssueTypeTask {
+			tasks = append(tasks, include.Title)
+		}
+	}
+
+	messages := make([]openai.ChatCompletionMessage, 0)
+
+	// 参考模板
+	messages = append(messages, openai.ChatCompletionMessage{
```
Comment (sfwn) [inline internal/pkg/ai-functions/functions/test-case/group.go:94]: I think it's better to define these messages in a yaml file.

File: internal/pkg/ai-functions/functions/test-case/group.go [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+	if len(completion.Choices) == 0 || completion.Choices[0].Message.Content == "" {
+		return errors.New("no idea")
+	}
+
+	contentBytes, _ := json.Marshal(completion.Choices[0].Message.Content)
+	logrus.Debugf("AI Gernerate completion.Choices[0].Message.Content=%s\n", string(contentBytes))
+
+	// "content" 示例
+	// "content": "'''\nDesired grouping results：\n模型触发\n定时触发\nwebhook触发\n通用工作流\n模型事件触发\n新增数据触发\n数据变更触发\n数据变更至条件范围内触发\n数据删除触发\n定时触发\n周期触发\n'''"
+	groupSources := strings.Split(string(contentBytes), "\\n")
```
Comment (sfwn) [inline internal/pkg/ai-functions/functions/test-case/group.go:192]: You can use function call to ensure grouped result in json slice format.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/apps/dop/endpoints/endpoints.go (hunk 1 above)
Comment (wangzhuzhen) [reply internal/apps/dop/endpoints/endpoints.go:282]: Get with request body?

File: api/proto/openapiv1/testplatform/testcase/testplatform_testcase.proto [hunk 9]
```diff
@@ -277,4 +278,20 @@ service testplatform_testcase {
             doc: "summary: 更新测试用例",
         };
     };
-}
+    rpc EXPORT_AI_GENERATED_TESTCASES (TestCaseExportRequest) returns (TestCaseExportResponse) {
```
[the remark this answers is not in the document]
Comment (wangzhuzhen) [reply api/proto/openapiv1/testplatform/testcase/testplatform_testcase.proto:281]: Ok

File: internal/apps/dop/endpoints/testcase.go (hunk 2 above)
Comment (wangzhuzhen) [reply internal/apps/dop/endpoints/testcase.go:445]: Ok

File: internal/core/openapi/legacy/api/apis/testplatform/testcase/export_ai_generated.go (hunk 4 above)
Comment (wangzhuzhen) [reply internal/core/openapi/legacy/api/apis/testplatform/testcase/export_ai_generated.go:24]: Ok

File: internal/pkg/ai-functions/functions/test-case/group.go (hunk 5 above)
Comment (wangzhuzhen) [reply internal/pkg/ai-functions/functions/test-case/group.go:38]: Ok

File: cmd/erda-server/bootstrap.yaml [hunk 10]
```diff
@@ -623,4 +623,6 @@ grpc-client@erda.core.monitor.settings:
 erda.core.monitor.settings-client:
 service-profile-overview:
 erda.app.ai-function:
-  openaiAddr: ${AI_PROXY_ADDR:http://ai-proxy:8081}
\ No newline at end of file
+  openaiAddr: ${AI_PROXY_ADDR:http://ai-proxy:8081}
+  modelIds:
+    chatgpt4: ${AI_PROXY_CHATGPT4_MODEL_ID:5426f79c-c43b-4cc5-9d6d-1598dc10f1be}
```
Comment (sfwn) [inline cmd/erda-server/bootstrap.yaml:628]: - not `chatgpt4`, just `gpt-4`
- do not leak model id here, inject using env

**Review by sfwn — CHANGES_REQUESTED** [review]

File: internal/pkg/ai-functions/handler/handler_ai_function.go [hunk 11]
```diff
@@ -51,19 +56,29 @@ func (h *AIFunction) Apply(ctx context.Context, req *pb.ApplyRequest) (pbValue *
...
+			h.Log.Warnf("config modelIds[\"chatgpt4\"] not set, use default")
+			xAIProxyModelId = DefaultChaGPT4ModelId
 		}
-
 	default:
 		err := errors.Errorf("AI function %s not support for apply", req.GetFunctionName())
 		return nil, HTTPError(err, http.StatusBadRequest)
 	}
 
+	results, err = f.Handler(ctx, factory, req, h.OpenaiURL, xAIProxyModelId)
```
Comment (sfwn) [inline internal/pkg/ai-functions/handler/handler_ai_function.go:77]: Great!

**Review by sfwn — APPROVED** [review]

**Pull request merged**